### ISSUE TITLE
TestPersistentVolumeClaimBinder flake

### ISSUE THREAD

**Issue by liggitt** [kubernetes/kubernetes#14284]
Seen in https://app.shippable.com/builds/56001b051a22100c00d15f6e running against https://github.com/kubernetes/kubernetes/pull/13835 (which had no API server changes)
```
W0921 15:45:04.798511   20405 controller.go:212] Resetting endpoints for master service "kubernetes" to &{{ } {kubernetes  default    0 0001-01-01 00:00:00 +0000 UTC   map[] map[]} [{[{172.17.5.229 }] [] [{ 6443 TCP}]}]}
E0921 15:45:04.858525   20405 repair.go:54] unable to persist the updated port allocations: 100: Key not found (/kubernetes.io) [159]
E0921 15:45:04.862195   20405 repair.go:69] unable to persist the updated service IP allocations: 100: Key not found (/kubernetes.io) [159]
E0921 15:45:05.041586   20405 persistentvolume_claim_binder_controller.go:132] PVClaimBinder could not add claim claim03: A volume match does not exist for persistent claim: claim03
--- FAIL: TestPersistentVolumeClaimBinder (0.55 seconds)
	persistent_volumes_test.go:109: Expected claim to be bound: &{TypeMeta:{Kind: APIVersion:} ObjectMeta:{Name:claim02 GenerateName: Namespace:default SelfLink:/api/v1/namespaces/default/persistentvolumeclaims/claim02 UID:ba75eb13-6077-11e5-a1f1-0242ac1105e5 ResourceVersion:166 Generation:0 CreationTimestamp:2015-09-21 15:45:05 +0000 UTC DeletionTimestamp: DeletionGracePeriodSeconds: Labels:map[] Annotations:map[]} Spec:{AccessModes:[ReadOnlyMany ReadWriteOnce ReadWriteMany] Resources:{Limits:map[] Requests:map[storage:{Amount:5000000000.000 Format:DecimalSI}]} VolumeName:} Status:{Phase:Pending AccessModes:[] Capacity:map[]}}
W0921 15:45:05.195558   20405 master.go:261] Network range for service cluster IPs is unspecified. Defaulting to 10.0.0.0/24.
I0921 15:45:05.195715   20405 master.go:287] Node port range unspecified. Defaulting to 30000-32767.
I0921 15:45:05.196348   20405 master.go:309] Will report 172.17.5.229 as public IP address.
```

cc @markturansky

**Comment by dchen1107** [comment]
cc/ @saad-ali

**Comment by dchen1107** [comment]
I saw a lot of shippable failure caused by this one. Another example:

https://app.shippable.com/builds/56006ef0b88a390c009e4e52

**Comment by saad-ali** [comment]
Dug into this. Looks like this happens because the test is not waiting for `claim02` become bound.

The test currently uses this code to determine if the claims are bound (`test/integration/persistent_volumes_test.go:89`):

```GO
	boundCount := 0
	expectedBoundCount := 2
	for {
		event := <-watch.ResultChan()
		claim := event.Object.(*api.PersistentVolumeClaim)
		if claim.Spec.VolumeName != "" {
			boundCount++
		}
		if boundCount == expectedBoundCount {
			break
		}
	}
```

It receives two events for `claim01` with `claim.Spec.VolumeName != ""`:
```
	persistent_volumes_test.go:99: TestPersistentVolumeClaimBinder: Recieved a bound: &{TypeMeta:{Kind: APIVersion:} ObjectMeta:{Name:claim01 GenerateName: Namespace:default SelfLink:/api/v1/namespaces/default/persistentvolumeclaims/claim01 UID:196b8331-60c6-11e5-be47-f0921cde18c1 ResourceVersion:166 Generation:0 CreationTimestamp:2015-09-21 18:06:05 -0700 PDT DeletionTimestamp:<nil> DeletionGracePeriodSeconds:<nil> Labels:map[] Annotations:map[]} Spec:{AccessModes:[ReadOnlyMany ReadWriteOnce] Resources:{Limits:map[] Requests:map[storage:{Amount:8000000000.000 Format:DecimalSI}]} VolumeName:gce003} Status:{Phase:Pending AccessModes:[] Capacity:map[]}}
	persistent_volumes_test.go:99: TestPersistentVolumeClaimBinder: Recieved a bound: &{TypeMeta:{Kind: APIVersion:} ObjectMeta:{Name:claim01 GenerateName: Namespace:default SelfLink:/api/v1/namespaces/default/persistentvolumeclaims/claim01 UID:196b8331-60c6-11e5-be47-f0921cde18c1 ResourceVersion:167 Generation:0 CreationTimestamp:2015-09-21 18:06:05 -0700 PDT DeletionTimestamp:<nil> DeletionGracePeriodSeconds:<nil> Labels:map[] Annotations:map[]} Spec:{AccessModes:[ReadOnlyMany ReadWriteOnce] Resources:{Limits:map[] Requests:map[storage:{Amount:8000000000.000 Format:DecimalSI}]} VolumeName:gce003} Status:{Phase:Bound AccessModes:[] Capacity:map[]}}
```

It then incorrectly assumes that both `claim01` and `claim02` are bound (`claim02` is still pending), and therefore the follow up check to verify the claims are bound fails for `claim02`.

I'll send out a PR to fix this.

**Comment by markturansky** [comment]
@saad-ali I thought about re-writing the waits, too, to be more like the newer integration tests I added recently.

I'll fix this up promptly and submit new tests for basic binding.

**Comment by saad-ali** [comment]
Re-opening. The test is flaky in a slightly different way now: See https://app.shippable.com/builds/5600dc6f28357b0c00e875d3

```
E0922 05:02:18.460290    9827 persistentvolume_claim_binder_controller.go:142] PVClaimBinder could not update claim claim03: A volume match does not exist for persistent claim: claim03
E0922 05:02:20.458632    9827 persistentvolume_claim_binder_controller.go:142] PVClaimBinder could not update claim claim03: A volume match does not exist for persistent claim: claim03
E0922 05:02:22.461488    9827 persistentvolume_claim_binder_controller.go:142] PVClaimBinder could not update claim claim03: A volume match does not exist for persistent claim: claim03
panic: test timed out after 2m0s

goroutine 2925 [running]:
runtime.panic(0x12b3140, 0xc20909c230)
	/root/.gvm/gos/go1.3/src/pkg/runtime/panic.c:279 +0xf5
testing.func·008()
	/root/.gvm/gos/go1.3/src/pkg/testing/testing.go:629 +0x12e
created by time.goFunc
	/root/.gvm/gos/go1.3/src/pkg/time/sleep.go:121 +0x55
...
goroutine 1403 [select]:

	/root/.gvm/gos/go1.3/src/pkg/net/http/transport.go:885 +0x483
created by net/http.(*Transport).dialConn
	/root/.gvm/gos/go1.3/src/pkg/net/http/transport.go:601 +0xd5f
net/http.(*persistConn).writeLoop(0xc2097834a0)
k8s.io/kubernetes/test/integration.TestPersistentVolumeClaimBinder(0xc208cc3440)
	/root/.gvm/gos/go1.3/src/pkg/testing/testing.go:422 +0x110
	/root/workspace/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/test/integration/persistent_volumes_test.go:93 +0x124b
testing.tRunner(0xc208cc3440, 0x2622ec0)
...
```

This indicates that the test `TestPersistentVolumeClaimBinder` timed-out waiting for `event := <-watch.ResultChan()` to return:
```GO
	// the binder will eventually catch up and set status on Claims
	watch, err := testClient.PersistentVolumeClaims(api.NamespaceDefault).Watch(labels.Everything(), fields.Everything(), "0")
...
	// Wait for claim01 and claim02 to become bound
	claim01Pending := true
	claim02Pending := true
	for claim01Pending || claim02Pending {
		event := <-watch.ResultChan()
		claim := event.Object.(*api.PersistentVolumeClaim)
		if claim.Spec.VolumeName != "" && claim.Status.Phase != "Bound" {
			if claim.Name == "claim01" {
				claim01Pending = false
			} else if claim.Name == "claim02" {
				claim02Pending = false
			}
		}
	}
```

1. This test is depending on watch `resourceVersion==0` behavior that will change with #13969.
2. We need to understand and fix the reason why `event := <-watch.ResultChan()` sometimes blocks for over 2 minutes.
  * Suggest replacing this check (watch and wait) with a periodic poll (get) to check status (and timeout/fail after some amount of time).

I'll take a crack at it in the morning if someone else doesn't get to it first (@pmorie? @markturansky?)

**Comment by markturansky** [comment]
@saad-ali the 2 newer integration tests I added (for Recycler and Deleter) both perform binding between a PV and PVC.   The original integration test (that one that's failing) only tests binding.  

We've got tons of unit tests that cover complex binding scenarios.  The simple scenario in the original integration test isn't that useful, IMHO, considering the newer ones also test the same thing.

Unless you object, I'd just rather drop the flaky test.  I believe we're sufficiently covered.

**Comment by soltysh** [comment]
One more here https://app.shippable.com/builds/56016eaf767f881200442642

**Comment by bprashanth** [comment]
https://app.shippable.com/builds/5601782585b6590c00f9a53f

**Comment by saad-ali** [comment]
Closed with #14356
